### DISCUSSION TITLE
[GS] Further control over GS story page formatting

### DISCUSSION THREAD

**Discussion by andythenorth** [OpenTTD/OpenTTD#10388] [Features]
It would be helpful to have _some_ more control over GS story page formatting.

I think that we want to be careful what we add.  Story pages are not supposed to provide sophisticated layout engines, and I don't think that's needed.

## Page flow control

Usually where the layout engine / UI is rudimentary, creative authors rely on things like text formatting.  Long history of that with shell, ASCII art etc.

Examples are things like dividers, which are often provided elsewhere as '-----------' or '========='.

<img width="771" alt="image" src="https://user-images.githubusercontent.com/1780327/213907997-1907cac5-f384-454e-a917-ff2803952abc.png">

Might be helpful to offer a little more control over that.  For example, text default aligns left.  Button elements can be somewhat floated left or right using StoryPageButtonFlags, but there's nothing currently to control text, for e.g alignment left/right/center.  

Might be a simple extension?  I know we have RTL language support to consider, but seems we'd just invert left/right for that?

## Element extension

Additionally / alternatively, we could extend the element types to include things like horizontal dividers?

## GSText as arg to text elements

It may be author error, but I can't get GSStoryPage.NewElement() to successfully use a GSText string for the 4th parameter (text to display).  Doxygen does _not_ mention this as supported, so it's not really a bug. 

Trying it does _not_ crash GS with a type error, which is what I was expecting, and the story page renders, but the contents of the GSText string are not displayed.  Maybe it's just getting an empty string or something in that case.

Seems that making GSText available in story page would be useful.  Or if it's already possible, updating doxygen comments to that effect.

## String control codes in elements

The usual string control codes like {GOLD} are not supported in strings for story page elements.  If they were supported by using a GSText instance, that might be a good solution, otherwise it might be helpful to have them directly for story pages.

**Comment by agentw4b** [comment]
Hello,
I would also like to be able to better influence the formatting of the buttons.
I wrote it here.... https://github.com/OpenTTD/OpenTTD/issues/12122

**Reply by agentw4b** [reply]
It is possible to place 3 buttons next to each other if the buttons have this order of parameters:
1. GSStoryPage.SPBF_FLOAT_LEFT;
2. GSStoryPage.SPBF_FLOAT_RIGHT;
3. GSStoryPage.SPBF_NONE;

So the position of the left and right buttons must be defined before the middle button. Any other order will result in the buttons failing to display correctly It's not very obvious from the documentation and it took me a few days to figure out the correct order of the buttons so that the 3 buttons are next to each other. :-)
Unfortunately, the formatting of the button width is kind of weird, I don't know what it's based on...
